### ISSUE TITLE
ICE in rustdoc when using `type_alias_impl_trait`

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/rust#73061]
### Code

```rust
//! crate a
#![feature(type_alias_impl_trait)]

pub trait Foo {
    type X: std::future::Future<Output = ()>;
    fn x(&self) -> Self::X;
}

pub struct F;

impl Foo for F {
    type X = impl std::future::Future<Output = ()>;
    fn x(&self) -> Self::X {
        async {}
    }
}
```

```rust
//! crate b, depends on crate a
pub struct Z;

impl a::Foo for Z {
    type X = <a::F as a::Foo>::X;
    fn x(&self) -> Self::X {
        a::F.x()
    }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.45.0-nightly (74e804683 2020-05-30)
binary: rustc
commit-hash: 74e80468347471779be6060d8d7d6d04e98e467f
commit-date: 2020-05-30
host: x86_64-unknown-linux-gnu
release: 1.45.0-nightly
LLVM version: 10.0
```

### Error output

Run `cargo doc --no-deps` in crate `b` (running it without `--no-deps` triggers https://github.com/rust-lang/rust/issues/65863 in `a`).

```
thread 'rustc' panicked at 'not implemented', src/librustdoc/clean/mod.rs:1311:40
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: std::panicking::begin_panic
  12: <rustc_middle::ty::AssocItem as rustdoc::clean::Clean<rustdoc::clean::types::Item>>::clean
  13: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
  14: rustdoc::clean::inline::build_impl
  15: rustdoc::passes::collect_trait_impls::collect_trait_impls
  16: rustc_middle::ty::context::tls::enter_global
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  19: rustc_driver::catch_fatal_errors
  20: rustdoc::main_options
  21: scoped_tls::ScopedKey<T>::set
  22: rustc_ast::attr::with_globals
```

**Comment by jonhoo** [comment]
The triggered assertion is here:
https://github.com/rust-lang/rust/blob/118b50524b79e565f017e08bce9b90a16c63634f/src/librustdoc/clean/mod.rs#L1311

**Comment by matthewjasper** [comment]
Fixed by #72080
